buildhub/main/models.py

```diff
@@ -111,6 +116,8 @@ def insert(cls, build, metadata=None, skip_validation=False, **kwargs):
         ):
             # If it returns something, it got created! Must inform Elasticsearch.
             send_to_elasticsearch(cls, build)
+            if settings.BQ_ENABLED:
+                send_to_bigquery(cls, build)
```

Comment:
What will happen if inserting to postgres/elasticsearch succeeds, but bigquery fails? Will the request fail overall, or will we end up in an inconsistent state? If the latter, do we have a way to recover from that (eg: by resyncing).

Reply:
We would end up in an inconsistent state. There are reasonable retry defaults with the BigQuery library, so if it does fail, it would mean there's a configuration error somewhere. For example, a change to the schema may cause data to be rejected from the BigQuery table if done in a backwards incompatible fashion.

There are a couple of different ways to recover from this state. One would be to create a new script that insert values starting from a particular timestamp, and then run a deduplicate query to bring the table up to date. Another would be to log/mark the builds that failed, and retry them at another time (or manually from a script).

Reply:
Fully agree - this is bound to happen at some point :)

Reply:
Thinking about this a bit more, it would be really, really nice if whatever process does the backfilling was quick enough that we could run it, eg: once an hour, to pick up anything that was missed. This would avoid the need for any manual intervention.

Do you have any idea how long a no-op run of this (where we check for existence of everything in BigQuery but don't need to insert anything) would take right now? If it takes too long this might not be possible without other changes.

Reply:
I wrote up an estimate of the size of the primary key set that would need to be used to filter out data that hasn't been picked up. I think it would make for a great follow-up to this PR, since the task is well defined and builds upon the work here. The logic is not complicated: it would materialize the set of keys locally from the BigQuery table, use it find the set difference, and then insert the straggling rows back into BigQuery.